From 9fd22ea533bbce44c763961ef54e25ab0f5be9c1 Mon Sep 17 00:00:00 2001
From: Mohd Faraz <androiabledroid@gmail.com>
Date: Sat, 7 Jan 2023 21:59:46 +0100
Subject: [PATCH] vold: dynamically choose fscrypt policy [2/2]

Change-Id: I076f5aecaa9102e2d14c80bd9a4daacaaa3b0e7b
Signed-off-by: Mohd Faraz <androiabledroid@gmail.com>
(cherry picked from commit b45786cb9e7c61f343202bc32664ea2a4206e8ff)

gvwifi port to android-14.1: kept the UserPolicies map type (s_ce/s_de_policies),
GET_ENCRYPTION_POLICY_EX without isFsKeyringSupported(), and Android.bp / vold_defaults
unchanged (libvold still uses the vold_defaults module type).

Change-Id: I71cef6d2c95071b97b5d49b0762d3e58f54a8910
---
 Decrypt.cpp          | 142 ++++++++++++++++---------------------------
 fscrypt_policy.cpp   | 110 ++++++++++++++++++---------------
 fscrypt_policy.h     |  56 ++++++++---------
 fscryptpolicyget.cpp |  15 +----
 4 files changed, 141 insertions(+), 182 deletions(-)

diff --git a/Decrypt.cpp b/Decrypt.cpp
index 0a5afb0..6d8cf50 100755
--- a/Decrypt.cpp
+++ b/Decrypt.cpp
@@ -38,7 +38,6 @@
 
 #include <dirent.h>
 #include <stdio.h>
-#include <stdint.h>
 #include <string.h>
 #include <sys/types.h>
 #include <fstream>
@@ -101,23 +100,13 @@ inline std::string hidlVec2String(const ::keystore::hidl_vec<uint8_t>& value) {
     return std::string(reinterpret_cast<const std::string::value_type*>(&value[0]), value.size());
 }
 
-static bool lookup_ref_key_internal(std::map<userid_t, UserPolicies> key_map, const uint8_t* policy, userid_t* user_id) {
-#ifdef USE_FSCRYPT_POLICY_V1
-	char policy_string_hex[FS_KEY_DESCRIPTOR_SIZE_HEX];
-	char key_map_hex[FS_KEY_DESCRIPTOR_SIZE_HEX];
-	bytes_to_hex(policy, FS_KEY_DESCRIPTOR_SIZE, policy_string_hex);
-#else
-	char policy_string_hex[FSCRYPT_KEY_IDENTIFIER_HEX_SIZE];
-	char key_map_hex[FSCRYPT_KEY_IDENTIFIER_HEX_SIZE];
-	bytes_to_hex(policy, FSCRYPT_KEY_IDENTIFIER_SIZE, policy_string_hex);
-#endif
+static bool lookup_ref_key_internal(std::map<userid_t, UserPolicies> key_map, const uint8_t* policy, uint8_t size, uint8_t hex_size, userid_t* user_id) {
+	char policy_string_hex[hex_size];
+	char key_map_hex[hex_size];
+	bytes_to_hex(policy, size, policy_string_hex);
 
     for (std::map<userid_t, UserPolicies>::iterator it=key_map.begin(); it!=key_map.end(); ++it) {
-#ifdef USE_FSCRYPT_POLICY_V1
-		bytes_to_hex(reinterpret_cast<const uint8_t*>(&it->second.internal.key_raw_ref[0]), FS_KEY_DESCRIPTOR_SIZE, key_map_hex);
-#else
-		bytes_to_hex(reinterpret_cast<const uint8_t*>(&it->second.internal.key_raw_ref[0]), FSCRYPT_KEY_IDENTIFIER_SIZE, key_map_hex);
-#endif
+		bytes_to_hex(reinterpret_cast<const uint8_t*>(&it->second.internal.key_raw_ref[0]), size, key_map_hex);
 		std::string key_map_hex_string = std::string(key_map_hex);
 		if (key_map_hex_string == policy_string_hex) {
             *user_id = it->first;
@@ -127,98 +116,69 @@ static bool lookup_ref_key_internal(std::map<userid_t, UserPolicies> key_map, co
     return false;
 }
 
-#ifdef USE_FSCRYPT_POLICY_V1
-extern "C" bool lookup_ref_key(fscrypt_policy_v1* fep, uint8_t* policy_type) {
-#else
-extern "C" bool lookup_ref_key(fscrypt_policy_v2* fep, uint8_t* policy_type) {
-#endif
+extern "C" bool lookup_ref_key(fscrypt_policy* fep, uint8_t* policy_type) {
 	userid_t user_id = 0;
 	std::string policy_type_string;
 
-#ifdef USE_FSCRYPT_POLICY_V1
-	char policy_hex[FS_KEY_DESCRIPTOR_SIZE_HEX];
-	bytes_to_hex(fep->master_key_descriptor, FS_KEY_DESCRIPTOR_SIZE, policy_hex);
-	if (std::strncmp((const char*)fep->master_key_descriptor, de_key_raw_ref.c_str(), FS_KEY_DESCRIPTOR_SIZE) == 0) {
-		policy_type_string = SYSTEM_DE_FSCRYPT_POLICY;
+	uint8_t *descriptor = get_policy_descriptor(fep);
+	uint8_t hex_size = get_policy_size(fep, true);
+	uint8_t size = get_policy_size(fep, false);
+	char policy_hex[hex_size];
+	bytes_to_hex(descriptor, size, policy_hex);
+	if (std::strncmp((const char*)descriptor, de_key_raw_ref.c_str(), size) == 0) {
+		policy_type_string = std::to_string(fep->version) + SYSTEM_DE_FSCRYPT_POLICY;
 		memcpy(policy_type, policy_type_string.data(), policy_type_string.size());
 		return true;
 	}
-    if (!lookup_ref_key_internal(s_de_policies, fep->master_key_descriptor, &user_id)) {
-        if (!lookup_ref_key_internal(s_ce_policies, fep->master_key_descriptor, &user_id)) {
-            return false;
-		} else {
-			policy_type_string = USER_CE_FSCRYPT_POLICY + std::to_string(user_id);
-		}
-    } else {
-			policy_type_string = USER_DE_FSCRYPT_POLICY + std::to_string(user_id);
-	}
-#else
-	char policy_hex[FSCRYPT_KEY_IDENTIFIER_HEX_SIZE];
-	bytes_to_hex(fep->master_key_identifier, FSCRYPT_KEY_IDENTIFIER_SIZE, policy_hex);
-	if (std::strncmp((const char*)fep->master_key_identifier, de_key_raw_ref.c_str(), FSCRYPT_KEY_IDENTIFIER_SIZE) == 0) {
-		policy_type_string = SYSTEM_DE_FSCRYPT_POLICY;
-		memcpy(policy_type, policy_type_string.data(), policy_type_string.size());
-		return true;
-	}
-    if (!lookup_ref_key_internal(s_de_policies, fep->master_key_identifier, &user_id)) {
-        if (!lookup_ref_key_internal(s_ce_policies, fep->master_key_identifier, &user_id)) {
-            return false;
-		} else {
-			policy_type_string = USER_CE_FSCRYPT_POLICY + std::to_string(user_id);
-		}
-    } else {
-			policy_type_string = USER_DE_FSCRYPT_POLICY + std::to_string(user_id);
-	}
-#endif
+	if (!lookup_ref_key_internal(s_de_policies, descriptor, size, hex_size, &user_id)) {
+		if (!lookup_ref_key_internal(s_ce_policies, descriptor, size, hex_size, &user_id)) return false;
+		else policy_type_string = std::to_string(fep->version) + USER_CE_FSCRYPT_POLICY + std::to_string(user_id);
+
+	} else policy_type_string = std::to_string(fep->version) + USER_DE_FSCRYPT_POLICY + std::to_string(user_id);
 
 	memcpy(policy_type, policy_type_string.data(), policy_type_string.size());
 	printf("storing policy type: %s\n", policy_type);
-    return true;
+	return true;
 }
 
-extern "C" bool lookup_ref_tar(const uint8_t* policy_type, uint8_t* policy) {
-	std::string policy_type_string = std::string((char *) policy_type);
-#ifdef USE_FSCRYPT_POLICY_V1
-	char policy_hex[FS_KEY_DESCRIPTOR_SIZE_HEX];
-	bytes_to_hex(policy_type, FS_KEY_DESCRIPTOR_SIZE, policy_hex);
-#else
-	char policy_hex[FSCRYPT_KEY_IDENTIFIER_HEX_SIZE];
-	bytes_to_hex(policy_type, FSCRYPT_KEY_IDENTIFIER_SIZE, policy_hex);
-#endif
-
-#ifdef USE_FSCRYPT_POLICY_V1
-	if (policy_type_string.substr(0,1) != FSCRYPT_V1) {
-#else
-	if (policy_type_string.substr(0,1) != FSCRYPT_V2) {
-#endif
-        printf("Unexpected version: %d\n", policy_type[0]);
-        return false;
-    }
-
+extern "C" bool lookup_ref_tar(fscrypt_policy *fep, uint8_t* policy) {
+	if (fep->version < FSCRYPT_POLICY_V1 || fep->version > FSCRYPT_POLICY_V2) {
+		printf("Unexpected version: %d\n", (int)fep->version);
+ 		return false;
+ 	}
+	uint8_t hex_size, size, *descriptor;
+	hex_size = get_policy_size(fep, true);
+	size = get_policy_size(fep, false);
+	descriptor = get_policy_descriptor(fep);
+	std::string policy_type_string = std::string((char *) descriptor);
+	char policy_hex[hex_size];
+	bytes_to_hex(descriptor, size, policy_hex);
 	if (policy_type_string.substr(1, 2) == SYSTEM_DE_KEY) {
-        memcpy(policy, de_key_raw_ref.data(), de_key_raw_ref.size());
-        return true;
-    }
+		memcpy(policy, de_key_raw_ref.data(), de_key_raw_ref.size());
+		return true;
+	}
 
-    std::string raw_ref;
+	std::string raw_ref;
 
 	if (policy_type_string.substr(1, 1) == USER_DE_KEY) {
 		userid_t user_id = std::stoi(policy_type_string.substr(3, 4).c_str());
-        if (lookup_key_ref(s_de_policies, user_id, &raw_ref)) {
-            memcpy(policy, raw_ref.data(), raw_ref.size());
-        } else
-            return false;
-    } else if (policy_type_string.substr(1, 1) == USER_CE_KEY) {
+		if (lookup_key_ref(s_de_policies, user_id, &raw_ref)) {
+			memcpy(policy, raw_ref.data(), raw_ref.size());
+		} else {
+			return false;
+		}
+	} else if (policy_type_string.substr(1, 1) == USER_CE_KEY) {
 		userid_t user_id = std::stoi(policy_type_string.substr(3, 4).c_str());
-        if (lookup_key_ref(s_ce_policies, user_id, &raw_ref)) {
-            memcpy(policy, raw_ref.data(), raw_ref.size());
-        } else
-            return false;
-    } else {
-        printf("unknown policy type: %s\n", policy_type);
-        return false;
-    }
-    return true;
+		if (lookup_key_ref(s_ce_policies, user_id, &raw_ref)) {
+			memcpy(policy, raw_ref.data(), raw_ref.size());
+		} else {
+			return false;
+		}
+	} else {
+		printf("unknown policy type: %s\n", descriptor);
+		return false;
+	}
+	return true;
 }
 
 extern "C" bool Decrypt_DE() {
diff --git a/fscrypt_policy.cpp b/fscrypt_policy.cpp
index 57dd544..5b2ce45 100755
--- a/fscrypt_policy.cpp
+++ b/fscrypt_policy.cpp
@@ -90,18 +90,14 @@ extern "C" bool fscrypt_set_mode() {
     return true;
 }
 
-#ifdef USE_FSCRYPT_POLICY_V1
-extern "C" bool fscrypt_policy_set_struct(const char *directory, const struct fscrypt_policy_v1 *fep) {
-#else
-extern "C" bool fscrypt_policy_set_struct(const char *directory, const struct fscrypt_policy_v2 *fep) {
-#endif
+extern "C" bool fscrypt_policy_set_struct(const char *directory, fscrypt_policy *fep) {
     int fd = open(directory, O_DIRECTORY | O_NOFOLLOW | O_CLOEXEC);
     if (fd == -1) {
         printf("failed to open %s\n", directory);
         PLOG(ERROR) << "Failed to open directory " << directory;
         return false;
     }
-    if (ioctl(fd, FS_IOC_SET_ENCRYPTION_POLICY, fep)) {
+    if (ioctl(fd, FS_IOC_SET_ENCRYPTION_POLICY, get_policy(fep))) {
         PLOG(ERROR) << "Failed to set encryption policy for " << directory;
         close(fd);
         return false;
@@ -110,21 +106,13 @@ extern "C" bool fscrypt_policy_set_struct(const char *directory, const struct fs
     return true;
 }
 
-#ifdef USE_FSCRYPT_POLICY_V1
-extern "C" bool fscrypt_policy_get_struct(const char *directory, struct fscrypt_policy_v1 *fep) {
-#else
-extern "C" bool fscrypt_policy_get_struct(const char *directory, struct fscrypt_policy_v2 *fep) {
-#endif
+extern "C" bool fscrypt_policy_get_struct(const char *directory, fscrypt_policy *fep) {
     int fd = open(directory, O_DIRECTORY | O_RDONLY | O_NOFOLLOW | O_CLOEXEC);
     if (fd == -1) {
         PLOG(ERROR) << "Failed to open directory " << directory;
         return false;
     }
-#ifdef USE_FSCRYPT_POLICY_V1
-    memset(fep, 0, sizeof(fscrypt_policy_v1));
-#else
-    memset(fep, 0, sizeof(fscrypt_policy_v2));
-#endif
+    memset(fep, 0, sizeof(fscrypt_policy));
     struct fscrypt_get_policy_ex_arg ex_policy = {0};
 
     ex_policy.policy_size = sizeof(ex_policy.policy);
@@ -133,49 +121,71 @@ extern "C" bool fscrypt_policy_get_struct(const char *directory, struct fscrypt_
         close(fd);
         return false;
     }
-#ifdef USE_FSCRYPT_POLICY_V1
-        memcpy(fep, &ex_policy.policy.v1, sizeof(ex_policy.policy.v1));
-#else
-        memcpy(fep, &ex_policy.policy.v2, sizeof(ex_policy.policy.v2));
-#endif
+    memcpy(fep, &ex_policy.policy, sizeof(ex_policy.policy));
     close(fd);
     return true;
 }
 
-#ifdef TW_INCLUDE_CRYPTO
-#ifdef USE_FSCRYPT_POLICY_V1
-extern "C" bool Get_Encryption_Policy(struct fscrypt_policy_v1 &policy, std::string path) {
-#else
-extern "C" bool Get_Encryption_Policy(struct fscrypt_policy_v2 &policy, std::string path) {
-#endif
-	if (!android::vold::pathExists(path)) {
-		LOGERR("Unable to find %s to get policy\n", path.c_str());
-		return false;
+extern "C" uint8_t* get_policy_descriptor(fscrypt_policy* fep) {
+	if (!fep) return NULL;
+	switch(fep->version) {
+		case FSCRYPT_POLICY_V1:
+			return fep->v1.master_key_descriptor;
+		case FSCRYPT_POLICY_V2:
+			return fep->v2.master_key_identifier;
+		default:
+			return NULL;
 	}
-	if (!fscrypt_policy_get_struct(path.c_str(), &policy)) {
-		LOGERR("No policy set for path %s\n", path.c_str());
-		return false;
+}
+
+extern "C" uint8_t get_policy_size(fscrypt_policy* fep, bool hex) {
+	if (!fep) return 0;
+	switch(fep->version) {
+		case FSCRYPT_POLICY_V1:
+			return hex ? FS_KEY_DESCRIPTOR_SIZE_HEX : FSCRYPT_KEY_DESCRIPTOR_SIZE;
+		case FSCRYPT_POLICY_V2:
+			return hex ? FSCRYPT_KEY_IDENTIFIER_SIZE_HEX : FSCRYPT_KEY_IDENTIFIER_SIZE;
 	}
-	return true;
+	return 0;
 }
 
-#ifdef USE_FSCRYPT_POLICY_V1
-extern "C" bool Set_Encryption_Policy(std::string path, struct fscrypt_policy_v1 &policy) {
-#else
-extern "C" bool Set_Encryption_Policy(std::string path, struct fscrypt_policy_v2 &policy) {
-#endif
-	if (!android::vold::pathExists(path)) {
-		LOGERR("unable to find %s to set policy\n", path.c_str());
-		return false;
+extern "C" void* get_policy(const fscrypt_policy *fep) {
+	if (!fep) return NULL;
+	switch(fep->version) {
+		case FSCRYPT_POLICY_V1:
+			return (void*)&(fep->v1);
+		case FSCRYPT_POLICY_V2:
+			return (void*)&(fep->v2);
 	}
-	uint8_t binary_policy[FS_KEY_DESCRIPTOR_SIZE];
-	char policy_hex[FSCRYPT_KEY_IDENTIFIER_HEX_SIZE];
-	bytes_to_hex(binary_policy, FS_KEY_DESCRIPTOR_SIZE, policy_hex);
-	if (!fscrypt_policy_set_struct(path.c_str(), &policy)) {
-		LOGERR("unable to set policy for path: %s\n", path.c_str());
-		return false;
+	return NULL;
+}
+
+extern "C" int fscrypt_policy_size(const fscrypt_policy *fep) {
+	switch (fep->version) {
+		case FSCRYPT_POLICY_V1:
+			return sizeof(fep->v1);
+		case FSCRYPT_POLICY_V2:
+			return sizeof(fep->v2);
 	}
-	return true;
+	return 0;
 }
-#endif
 
+void get_policy_content(fscrypt_policy* fep, char* content) {
+	if (!fep || !content) return;
+	switch(fep->version) {
+		case FSCRYPT_POLICY_V1:
+			sprintf(content, "%i %i %i %i %s", (int)fep->v1.version,
+				(int)fep->v1.contents_encryption_mode,
+				(int)fep->v1.filenames_encryption_mode,
+				(int)fep->v1.flags,
+				fep->v1.master_key_descriptor);
+			break;
+		case FSCRYPT_POLICY_V2:
+			sprintf(content, "%i %i %i %i %s", (int)fep->v2.version,
+				(int)fep->v2.contents_encryption_mode,
+				(int)fep->v2.filenames_encryption_mode,
+				(int)fep->v2.flags,
+				fep->v2.master_key_identifier);
+			break;
+	}
+}
diff --git a/fscrypt_policy.h b/fscrypt_policy.h
index 5714790..7b6e07a 100755
--- a/fscrypt_policy.h
+++ b/fscrypt_policy.h
@@ -19,26 +19,19 @@
 
 #include <sys/cdefs.h>
 #include <stdbool.h>
+#include <stdint.h>
 #include <cutils/multiuser.h>
 #include <linux/fs.h>
 
 __BEGIN_DECLS
 
 #define FS_KEY_DESCRIPTOR_SIZE_HEX (2 * FS_KEY_DESCRIPTOR_SIZE + 1)
-#define FSCRYPT_KEY_IDENTIFIER_HEX_SIZE ((2 * FSCRYPT_KEY_IDENTIFIER_SIZE) + 1)
-
-#ifdef USE_FSCRYPT_POLICY_V1
-#define USER_CE_FSCRYPT_POLICY           "0CE"
-#define USER_DE_FSCRYPT_POLICY           "0DE"
-#define SYSTEM_DE_FSCRYPT_POLICY         "0DK"
-#else
-#define USER_CE_FSCRYPT_POLICY           "2CE"
-#define USER_DE_FSCRYPT_POLICY           "2DE"
-#define SYSTEM_DE_FSCRYPT_POLICY         "2DK"
-#endif
+#define FSCRYPT_KEY_IDENTIFIER_SIZE_HEX ((2 * FSCRYPT_KEY_IDENTIFIER_SIZE) + 1)
+
+#define USER_CE_FSCRYPT_POLICY           "CE"
+#define USER_DE_FSCRYPT_POLICY           "DE"
+#define SYSTEM_DE_FSCRYPT_POLICY         "DK"
 
-#define FSCRYPT_V1                        "0"
-#define FSCRYPT_V2                        "2"
 #define SYSTEM_DE_KEY                     "DK"
 #define USER_CE_KEY                       "C"
 #define USER_DE_KEY                       "D"
@@ -61,27 +54,32 @@ __BEGIN_DECLS
 
 bool fscrypt_set_mode();
 
-#ifdef USE_FSCRYPT_POLICY_V1
-bool lookup_ref_key(struct fscrypt_policy_v1 *fep, uint8_t* policy_type);
-#else
-bool lookup_ref_key(struct fscrypt_policy_v2 *fep, uint8_t* policy_type);
-#endif
+#undef fscrypt_policy
+typedef union fscrypt_policy {
+	uint8_t version;
+	struct fscrypt_policy_v1 v1;
+	struct fscrypt_policy_v2 v2;
+} fscrypt_policy;
 
-bool lookup_ref_tar(const uint8_t *policy_type, uint8_t *policy);
+bool lookup_ref_key(fscrypt_policy *fep, uint8_t* policy_type);
 
-#ifdef USE_FSCRYPT_POLICY_V1
-bool fscrypt_policy_get_struct(const char *directory, struct fscrypt_policy_v1  *fep);
-#else
-bool fscrypt_policy_get_struct(const char *directory, struct fscrypt_policy_v2  *fep);
-#endif
+bool lookup_ref_tar(fscrypt_policy *fep, uint8_t *policy);
 
-#ifdef USE_FSCRYPT_POLICY_V1
-bool fscrypt_policy_set_struct(const char *directory, const struct fscrypt_policy_v1  *fep);
-#else
-bool fscrypt_policy_set_struct(const char *directory, const struct fscrypt_policy_v2  *fep);
-#endif
+bool fscrypt_policy_get_struct(const char *directory, fscrypt_policy *fep);
+
+bool fscrypt_policy_set_struct(const char *directory, fscrypt_policy  *fep);
 
 void bytes_to_hex(const uint8_t *bytes, size_t num_bytes, char *hex);
+
+uint8_t* get_policy_descriptor(fscrypt_policy* fep);
+
+void* get_policy(const fscrypt_policy* fep);
+
+uint8_t get_policy_size(fscrypt_policy* fep, bool hex);
+
+int fscrypt_policy_size(const fscrypt_policy *policy);
+
+void get_policy_content(fscrypt_policy* fep, char* content);
 __END_DECLS
 
 #endif // _FS_CRYPT_H_
diff --git a/fscryptpolicyget.cpp b/fscryptpolicyget.cpp
index 189a428..f9b0004 100755
--- a/fscryptpolicyget.cpp
+++ b/fscryptpolicyget.cpp
@@ -25,19 +25,10 @@ int main(int argc, char *argv[]) {
 		printf("Must specify a path\n");
 		return -1;
 	} else  {
-#ifdef USE_FSCRYPT_POLICY_V1
-		fscrypt_policy_v1 fep;
-#else
-		fscrypt_policy_v2 fep;
-#endif
+		fscrypt_policy fep;
 		if (fscrypt_policy_get_struct(argv[1], &fep)) {
-#ifdef USE_FSCRYPT_POLICY_V1
-			char policy_hex[FS_KEY_DESCRIPTOR_SIZE_HEX];
-			bytes_to_hex(fep.master_key_descriptor, FS_KEY_DESCRIPTOR_SIZE, policy_hex);
-#else
-			char policy_hex[FSCRYPT_KEY_IDENTIFIER_HEX_SIZE];
-			bytes_to_hex(fep.master_key_identifier, FSCRYPT_KEY_IDENTIFIER_SIZE, policy_hex);
-#endif
+			char policy_hex[get_policy_size(&fep, true)];
+			bytes_to_hex(get_policy_descriptor(&fep), get_policy_size(&fep, false), policy_hex);
 			printf("%s\n", policy_hex);
 		} else {
 			printf("No policy set\n");
